apm-agent-plugins/apm-opentelemetry/apm-opentelemetry-metricsdk-plugin/src/test/java/co/elastic/apm/agent/otelmetricsdk/AbstractOtelMetricsTest.java

```diff
@@ -0,0 +1,624 @@
+            .first()
+            .containsValueMetric("double_counter", 0.0)
+            .containsValueMetric("long_counter", 0)
+            .hasMetricsCount(2);
+
+        doubleCnt.add(-150);
+        doubleCnt.add(49.5);
+        longCnt.add(-10);
+        longCnt.add(1);
+
+        resetReporterAndFlushMetrics();
+        assertThatMetricSets(reporter.getBytes())
+            .hasMetricsetCount(1)
+            .first()
+            .containsValueMetric("double_counter", -100.5)
+            .containsValueMetric("long_counter", -9)
+            .hasMetricsCount(2);
+
+        resetReporterAndFlushMetrics();
+        assertThatMetricSets(reporter.getBytes())
+            .hasMetricsetCount(1)
+            .first()
+            .containsValueMetric("double_counter", -100.5)
+            .containsValueMetric("long_counter", -9)
+            .hasMetricsCount(2);
+    }
+
+
+    @Test
+    public void testHistogram() {
+        MetricsConfiguration metricsConfig = config.getConfig(MetricsConfiguration.class);
+        doReturn(List.of(1.0, 5.0)).when(metricsConfig).getCustomMetricsHistogramBoundaries();
+
+        Meter testMeter = createMeter("test");
+        DoubleHistogram doubleHisto = testMeter.histogramBuilder("double_histo").build();
+        LongHistogram longHisto = testMeter.histogramBuilder("long_histo").ofLongs().build();
+
+        doubleHisto.record(0.5);
+        doubleHisto.record(1.5);
+        doubleHisto.record(1.5);
+        doubleHisto.record(5.5);
+        doubleHisto.record(5.5);
+        doubleHisto.record(5.5);
+
+        longHisto.record(0);
+        longHisto.record(2);
+        longHisto.record(2);
+        longHisto.record(6);
+        longHisto.record(6);
+        longHisto.record(6);
+
+        resetReporterAndFlushMetrics();
+        assertThatMetricSets(reporter.getBytes())
+            .hasMetricsetCount(1)
+            .first()
+            .containsHistogramMetric("double_histo", List.of(0.5, 3.0, 5.0), List.of(1L, 2L, 3L))
+            .containsHistogramMetric("long_histo", List.of(0.5, 3.0, 5.0), List.of(1L, 2L, 3L))
+            .hasMetricsCount(2);
+
+        //make sure only delta is reported and empty buckets are omitted
+        doubleHisto.record(1.5);
+        longHisto.record(2);
+
+        resetReporterAndFlushMetrics();
+        assertThatMetricSets(reporter.getBytes())
+            .hasMetricsetCount(1)
+            .first()
+            .containsHistogramMetric("double_histo", List.of(3.0), List.of(1L))
+            .containsHistogramMetric("long_histo", List.of(3.0), List.of(1L))
+            .hasMetricsCount(2);
+
+        //empty histograms must not be exported
+        resetReporterAndFlushMetrics();
+        assertThatMetricSets(reporter.getBytes())
+            .hasMetricsetCount(0);
+    }
+
+
+    /**
+     * Smoke test which just checks that nothing breaks with the default histogram buckets.
+     */
+    @Test
+    public void testDefaultHistogramBuckets() {
```

Comment:
:star_struck: That's a very good check to ensure the values we have in the spec make sense, I wonder if we could also do the same to estimate the relative accuracy of the histogram.

Reply:
I don't think this would give any benefit here, as the upper bound for the relative error is defined by how the boundaries are defined in the spec: I used a nexponential histogram with base `sqrt(2)=~1.41`, which means we have a worst case relative accuracy of 41% for values which lie within the maximum / minimum of the boundaries. This is still not great, in the future we should switch to either T-Digests or HDRHistogram...